### PR TITLE
Add ApiSpecFetcher for Fetching and Comparing API Specifications

### PR DESCRIPTION
**Pull request by junweid62**
### Description

- Introduced a new `ApiSpecFetcher` utility to programmatically fetch and compare OpenAPI specifications.

- Supports comparing required input fields between an OpenAPI spec and predefined enum values.

- The current implementation serves as a foundational mechanism, with the comparison logic successfully integrated into the `Register Agent` workflow.

- Future PRs will extend this mechanism to other workflow steps, ensuring consistent validation across the API specification.

### Related Issues
Part of #651

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/flow-framework/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/flowframework/util/ApiSpecFetcherTests.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    private ApiSpecFetcher apiSpecFetcher;
+    private URI apiSpecUri;
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        this.apiSpecFetcher = new ApiSpecFetcher(); // Initialize your fetcher
+        this.apiSpecUri = new URI(
+            "https://raw.githubusercontent.com/opensearch-project/opensearch-api-specification/refs/heads/main/spec/namespaces/ml.yaml"
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/flowframework/util/ApiSpecFetcherTests.java:32]: Let's make this a constant.

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    /**
+     * Parses the OpenAPI specification directly from the URI.
+     *
+     * @param apiSpecUri URI to the API specification (can be file path or web URI).
+     * @return Parsed OpenAPI object.
+     * @throws ApiSpecParseException If parsing fails.
+     */
+    public OpenAPI fetchApiSpec(URI apiSpecUri) {
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:55]: Looking at the test classes, we always create this URI from a String, then you spend a lot of time undoing the string in this method.  Do we need a URI argument here or can we just take a string?

**Review by dbwiddis — COMMENTED** [review]
Looks great!  A few suggestions.

File: src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.exception;
+
+import org.opensearch.OpenSearchException;
+
+/**
+ * Custom exception to be thrown when an error occurs during the parsing of an API specification.
+ */
+public class ApiSpecParseException extends OpenSearchException {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java:16]: We have FlowFrameworkException which extends from `OpenSearchException`, should we utilize the same class for `ApiSpecParseException`?

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import io.swagger.v3.oas.models.parameters.RequestBody;
+import io.swagger.v3.parser.OpenAPIV3Parser;
+import io.swagger.v3.parser.core.models.ParseOptions;
+import io.swagger.v3.parser.core.models.SwaggerParseResult;
+
+/**
+ * Utility class for fetching and parsing OpenAPI specifications.
+ */
+public class ApiSpecFetcher {
+    private static final Logger LOGGER = LogManager.getLogger(ApiSpecFetcher.class);
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:35]:
```suggestion
    private static final Logger logger = LogManager.getLogger(ApiSpecFetcher.class);
```

File: src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java [hunk 5]
```diff
@@ -150,4 +154,21 @@ public void testRegisterAgentFailure() throws IOException {
...
     }
+
+    public void testApiSpecRegisterAgentInputParamComparison() throws Exception {
+        List<String> requiredEnumParams = WorkflowStepFactory.WorkflowSteps.REGISTER_AGENT.inputs();
+
+        ApiSpecFetcher apiSpecFetcher = new ApiSpecFetcher();
+        boolean isMatch = apiSpecFetcher.compareRequiredFields(
+            requiredEnumParams,
+            new URI(
+                "https://raw.githubusercontent.com/opensearch-project/opensearch-api-specification/refs/heads/main/spec/namespaces/ml.yaml"
```
Comment (owaiskazi19) [inline src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java:165]: We can have a constant for it?

File: src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java [hunk 6]
```diff
@@ -150,4 +154,21 @@ public void testRegisterAgentFailure() throws IOException {
         assertTrue(ex.getCause() instanceof FlowFrameworkException);
         assertEquals("Failed to register the agent", ex.getCause().getMessage());
     }
+
+    public void testApiSpecRegisterAgentInputParamComparison() throws Exception {
```
Comment (owaiskazi19) [inline src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java:158]: Can we have a success test as well?

**Review by owaiskazi19 — COMMENTED** [review]
Great start!

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java (hunk 2 above)
Comment (junweid62) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:55]: I think string would be good, I simply followed the pattern in opensearch-java before.

File: src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (junweid62) [reply src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java:18]: I see the point about having `FlowFrameworkException`, but I believe `ApiSpecParseException` serves a distinct purpose by clearly indicating issues related to API spec parsing. Keeping it separate helps with better error categorization and makes it easier to track and handle specific parsing-related problems.

[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java:18]: @owaiskazi19 I agree with @junweid62 that there's room for a separate exception type here.  We primarily use FFE during operation, but this exception is almost exclusively used in tests.

That said, all the code here is in `main` but primarily meant for `test` so we _could_ move it there; even if just temporarily until we decide that we want to create an `ApiSpecStep` for our workflows.   WDYT?

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java (hunk 2 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java]: +1 for copying existing code, but they probably have other external API requirements/integrations expecting that parameter.  I don't mind leaving it as-is in the interest of keeping code consistent (although a comment pointing to the source of the code for later cross-referencing would be nice).

File: build.gradle [hunk 7]
```diff
@@ -33,7 +33,11 @@ buildscript {
         bwcFilePath = "src/test/resources/org/opensearch/flowframework/bwc/"
         bwcFlowFrameworkPath = bwcFilePath + "flowframework/"
 
-        isSameMajorVersion = opensearch_version.split("\\.")[0] == bwcVersionShort.split("\\.")[0]
+        isSameMajorVersion = opensearch_version.split("\\.")[0] == bwcVersionShort.split("\\.")
```
Comment (dbwiddis) [inline build.gradle:36]: Is this change intentional?

File: src/main/java/org/opensearch/flowframework/common/CommonValue.java [hunk 8]
```diff
@@ -233,4 +233,12 @@ private CommonValue() {}
     public static final String CREATE_INGEST_PIPELINE_MODEL_ID = "create_ingest_pipeline.model_id";
     /** The field name for reindex source index substitution */
     public static final String REINDEX_SOURCE_INDEX = "reindex.source_index";
+
+    /*
+     * Constants associated with API specification
+     */
+
+    public static final String ML_COMMONS_API_SPEC_YAML_URI =
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/common/CommonValue.java:241]: Add a `/**` comment for the constant to make it part of the javadoc.  You can also put it on one line.  Not sure it needs its own category yet but won't hurt to leave that there for now.

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    private static final ParseOptions PARSE_OPTIONS = new ParseOptions();
+    private static final OpenAPIV3Parser PARSER = new OpenAPIV3Parser();
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:37]: I think you changed the pares options name here (which was ok before) instead of the actual parser name (my suggestion) :)

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        return false;
+    }
+
+    private static Content getContent(RestRequest.Method method, PathItem pathItem) throws IllegalArgumentException, ApiSpecParseException {
+        Operation operation = switch (method) {
+            case RestRequest.Method.POST -> pathItem.getPost();
+            case RestRequest.Method.GET -> pathItem.getGet();
+            case RestRequest.Method.PUT -> pathItem.getPut();
+            case RestRequest.Method.DELETE -> pathItem.getDelete();
+            default -> throw new IllegalArgumentException("Unsupported HTTP method: " + method);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:96]: CodeCov indicates you haven't added tests for this failure case (and 2 below).

File: src/test/java/org/opensearch/flowframework/exception/ApiSpecParseExceptionTests.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    public void testApiSpecParseException() {
+        ApiSpecParseException exception = new ApiSpecParseException("API spec parsing failed");
+        assertTrue(true);
+        assertEquals("API spec parsing failed", exception.getMessage());
+    }
+
+    public void testApiSpecParseExceptionWithCause() {
+        Throwable cause = new RuntimeException("Underlying issue");
+        ApiSpecParseException exception = new ApiSpecParseException("API spec parsing failed", cause);
+        assertTrue(true);
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/flowframework/exception/ApiSpecParseExceptionTests.java:27]: image

**Review by dbwiddis — COMMENTED** [review]
A few more tweaks.

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java (hunk 10 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:96]: One more note:  we will have to backport this to 2.x which still supports JDK 11.   While I love the new functionality here, we may want to stick with old-style case statements.

File: src/test/java/org/opensearch/flowframework/exception/ApiSpecParseExceptionTests.java (hunk 11 above)
Comment (junweid62) [reply src/test/java/org/opensearch/flowframework/exception/ApiSpecParseExceptionTests.java:27]: Haha, I just realized how funny this is. This will be the last time I use a code refactor suggestion without thinking it through.
Screenshot 2024-10-07 at 1 48 19 PM

File: src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java (hunk 6 above)
Comment (owaiskazi19) [reply src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java:158]: Why do we have the below then?
```
 assertTrue("API spec input params do match enum required params", isMatch);
```

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+            List<String> requiredApiParams = schema.getRequired();
+            if (requiredApiParams != null && !requiredApiParams.isEmpty()) {
+                return new HashSet<>(requiredEnumParams).equals(new HashSet<>(requiredApiParams));
+            }
+        }
+        return false;
+    }
+
+    private static Content getContent(RestRequest.Method method, PathItem pathItem) throws IllegalArgumentException, ApiSpecParseException {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:90]: We can have a test for this method if we are not testing it at other places

File: src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java (hunk 6 above)
Comment (junweid62) [reply src/test/java/org/opensearch/flowframework/workflow/RegisterAgentTests.java:158]: sure , we only need to verify the isMatch is true, will change

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java (hunk 12 above)
Comment (junweid62) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:90]: I add more test included this private method.

File: src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/exception/ApiSpecParseException.java:18]: Agreed, it's fine in main and we will probably use it in the future.

File: src/test/java/org/opensearch/flowframework/util/ApiSpecFetcherTests.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+        boolean comparisonResult = ApiSpecFetcher.compareRequiredFields(wrongRequiredParams, ML_COMMONS_API_SPEC_YAML_URI, path, method);
+
+        assertFalse("The required fields should not match for incorrect input", comparisonResult);
+    }
+
+    public void testCompareRequiredFieldsThrowsException() throws Exception {
+        String invalidUri = "http://invalid-url.com/fail.yaml";
+        String path = "/_plugins/_ml/agents/_register";
+        RestRequest.Method method = RestRequest.Method.POST;
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/flowframework/util/ApiSpecFetcherTests.java:76]: If you change this to `PUT` you'll cover another line in your switch/case.

File: src/test/java/org/opensearch/flowframework/util/ApiSpecFetcherTests.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.util;
+
+import org.opensearch.flowframework.exception.ApiSpecParseException;
+import org.opensearch.rest.RestRequest;
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/flowframework/util/ApiSpecFetcherTests.java:12]: (Style) If you import `RestRequest.Method` you'll have less verbose class definitions later.

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java:37]: This comment wasn't addressed.

**Review by dbwiddis — APPROVED** [review]
Approving with a few last items to clean up

File: build.gradle [hunk 15]
```diff
@@ -179,6 +183,15 @@ dependencies {
...
     implementation "com.google.code.gson:gson:2.11.0"
+    // Swagger-Parser dependencies for API consistency tests
+    implementation "io.swagger.core.v3:swagger-models:${swaggerCoreVersion}"
+    implementation "io.swagger.core.v3:swagger-core:${swaggerCoreVersion}"
+    implementation "io.swagger.parser.v3:swagger-parser-core:${swaggerVersion}"
+    implementation "io.swagger.parser.v3:swagger-parser:${swaggerVersion}"
+    implementation "io.swagger.parser.v3:swagger-parser-v3:${swaggerVersion}"
+    implementation "com.fasterxml.jackson.core:jackson-databind:${jacksonVersion}"
+    implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:${jacksonVersion}"
+    implementation "com.fasterxml.jackson.core:jackson-annotations:${jacksonVersion}"
```
Comment (owaiskazi19) [inline build.gradle:194]: jackson dependencies are forced on L205. L192-194 should be moved under configurations as well

Comment (junweid62) [reply build.gradle:194]: Thank you for your suggestion. I tried moving the Jackson dependencies under configurations with force, but it seems that doing so doesn't properly import the dependencies. The force directive only enforces specific versions if the dependencies are already present, but it doesn't add them to the project on its own. Therefore, the implementation lines in dependencies are still necessary to ensure the Jackson libraries are actually included.

Comment (owaiskazi19) [reply build.gradle:194]: > only enforces specific versions 

And you are using the same version, right? 2.18.0. Any error you are seeing?

Comment (junweid62) [reply build.gradle:194]: Yes, I am using version 2.18.0 as specified in the force directive. The error I'm encountering is:

`java.lang.ClassNotFoundException: com.fasterxml.jackson.databind.Module`

It seems like the jackson-databind dependency isn't being imported properly, which is likely causing the issue. Even though the version is forced, the dependency itself doesn't get added unless it's explicitly declared. That's why I still need the implementation lines in the dependencies block to ensure that these libraries are available in the project.

Comment (owaiskazi19) [reply build.gradle:194]: I see. Let's discuss more on this. Approving the PR

**Review by owaiskazi19 — APPROVED** [review]

File: src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/ApiSpecFetcher.java]: It is now :)
